src/files.ts

```diff
@@ -187,22 +193,61 @@ export async function collectIgnoreRules(
     // Check if symlink and exclude if requested
     if (!fileStats || (fileStats.isSymbolicLink() && !symlinksEnabled) || fileStats.isFile()) return [];
 
-    // Find ignore files inside this directory
-    const localIgnoreFiles = await fg(
+    // Find .snyk and .[*]ignore files inside this directory.
+    const allIgnoredFiles = await fg(
       IGNORE_FILES_NAMES.map(i => `*${i}`),
       {
         ...fgOptions,
         cwd: folder,
         followSymbolicLinks: symlinksEnabled,
       },
     );
-    // Read ignore files and merge new patterns
-    return union(...localIgnoreFiles.map(parseFileIgnores));
+
+    // Parse rules from all .snyk files inside this directory.
+    const snykFiles = allIgnoredFiles.filter(f => f.endsWith(DOTSNYK_FILENAME));
+    const snykExcludeRules = union(...snykFiles.map(parseFileIgnores));
+
+    // Parse rules from all .[*]ignore files inside this directory.
+    const ignoreFiles = allIgnoredFiles.filter(
+      // Exclude .snyk files (parsed separately above) and files in paths excluded by .snyk files.
+      f => !f.endsWith(DOTSNYK_FILENAME) && multimatch(f, snykExcludeRules).length === 0,
+    );
+
+    return parseIgnoreRulesFromFiles(ignoreFiles, snykExcludeRules);
   });
+
   const customRules = await Promise.all(tasks);
+
   return union(fileIgnores, ...customRules);
 }
 
+/**
+ * Parse and collect ignore rules from .[*]ignore files.
+ * Removes redundant rules for efficiency.
+ * Removes negative rules that are superseded by .snyk exclude rules.
+ */
+function parseIgnoreRulesFromFiles(ignoreFiles: string[], snykExcludeRules: string[]): string[] {
+  // Read ignore files and merge patterns.
+  const localIgnoreRules = union(...ignoreFiles.map(parseFileIgnores));
+
+  const rulesAfterProcessing: string[] = [];
+  return union(localIgnoreRules, snykExcludeRules).filter((rule: string) => {
+    // Identify and remove negative rules that are superseded by .snyk exclude rules.
+    // i.e. if a directory "foo" is ignored by a .snyk exclude rule, then a negative match
+    // for "!foo" from another ignore file is superseded and needs to be removed.
+    const negativeRuleIsSupersededBySnykExcludes =
+      rule.startsWith('!') && snykExcludeRules.some(existingRule => multimatch(rule.slice(1), existingRule).length > 0);
```

Comment:
q: Is it really correct to match rules against each other like this? I'm thinking something like this:

Directory tree:
```
root/
  sub/
    exclude.txt
```

Superseding rule A (e.g. from `.snyk`): `root/**/exclude.txt`
Superseded rule B (e.g. from `.gitignore`): `!root/sub`


If I understand the logic here correctly, for rule B we would have:
- `negativeRuleIsSupersededBySnykExcludes = false`
- `ruleIsRedundant = false`
- `ruleIsValid = true`

And therefore rule B would again incorrectly override the exclusion from rule A.

I have pushed an example test for what I mean [here](https://github.com/snyk/code-client/commit/740169a3f7ae5459a09be668de1919934f5b59e5).

Reply:
Good catch, will look into this case, however your commit is adding a new test (marked broken, but which passes), and removes the snapshot contents from the previous test which fails. Apparently `matchInlineSnapshot` does not like the existing snapshots with dynamic path prefixes and just removes them.

Reply:
I was just pushing it as an example to make it clearer. The test passes, but the output is incorrect, right?

Reply:
Ah, I see. So the rule collection seems correct to me, as we shouldn't drop the rule for `!root/sub` entirely, but because of our 2-pass implementation with negative patterns second we'll still see that `/root/sub/exclude.txt` in the bundle files.

One thing that comes to mind is feeding the positive ignore rules back into the `negativeSearcher` fg call, which should ensure that ignored files are really ignored and not surfaced by the second pass, but unsure what the implications are then.

Will think about this some more.
cc @shinin-light , if you have any thoughts 🙏

Reply:
> feeding the positive ignore rules back into the negativeSearcher fg call, which should ensure that ignored files are really ignored and not surfaced by the second pass

I'm pretty sure I've tried something like that already and it failed. I remember playing around both filters and positive/negative ignores both in the 1-pass and 2-pass scenarios since `fg` allows you to have both patterns and exclusions, but in the end I remember stopping after successfully applying "negated filters" as "excludes" in the second pass. Feel free to experiment more

Reply:
Hmmm, yeah, we'd probably have to limit the ignores for the second pass to only the .snyk rules, not all of them, otherwise negative .gitignore matchers would still not work...

Reply:
I think I found a solution, that also works for the `TODO` case we had with the `/not/ignored/this_should_be_ignored.jsx` file:
On the second pass with the negative patterns we provide the positive ignore rules that also match one of our supported files patterns.
This should ensure files under paths which have been ignored and then overridden by a negative rule are surfaced, but specific files we support that have been marked ignored are still ignored in the second pass.
It's not perfect, but IMO it improves correctness a lot and should minimise edge cases we have.

Reply:
And a breaking case:
- A `.snyk` file excluding `/gitignore-unexcluded/sub`, and
- A `.gitignore` file with a negative rule for `!/gitignore-unexcluded`
Will still surface files from `/gitignore-unexcluded/sub`, so not correct.

After discussing with @strassl-snyk it looks like our best option (short of writing our own lib implementation) is to only feed in `.snyk` rules to the second pass `ignore`.
Some diagrams that may help understanding these flows [here](https://miro.com/app/board/uXjVMnCevuQ=/?share_link_id=301693422540).
Will be working on the "Other proposal" changes now which should address all edge cases so far as well as the breaking case described above.

Reply:
Updated implementation:
- Rule filtering was removed, we now parse all `.snyk` files, and then parse all other ignore files from paths that were not excluded.
- File ignores are now returned as a `FilePolicies` object with separate `excludes` and `ignores`, so in the file collection step we can apply all positive ignores (combined), and only the `.snyk` excludes in the second pass for negative patterns.

The case described above is fixed, but the previous case with the `/not/ignored/this_should_be_ignored.jsx` file is back 😢 
So we still don't correctly apply negative patterns, but we do ensure that `.snyk` excludes always take precedence and that negative patterns will never override excludes.